LayoutTests/pointer-lock/pending-locked-element-removed-from-dom.html

```diff
@@ -0,0 +1,48 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+<script src="../http/tests/resources/js-test-pre.js"></script>
+<script src="../http/tests/resources/pointer-lock/pointer-lock-test-harness.js"></script>
+</head>
+<body>
+<div>
+  <div id="target"></div>
+</div>
+<script>
+    description("Test removing an element with a pending pointer lock cancels the lock.")
+    window.jsTestIsAsync = true;
+    window.internals.settings.setPointerLockOptionsEnabled(true);
+
+    targetDiv = document.getElementById("target");
+
+    todo = [
+        function () {
+            // doNextStepWithUserGesture() has the side effect on the TestRunner that any 
+            // normally asynchronous operations such as requestPointerLock are actually completed synchronously.
+            // (see UseFullySynchronousModeForTesting)
+            //
+            // For this test, we don't want that, we want the normal behavior we would have in the browser
+            // where there is a pending lock period between requestPointerLock and when the pointer is
+            // actually locked.
+            //
+            // Use setTimeout to turn the event loop while still maintaining user interaction to let
+            // requestPointerLock be allowed, but have it run asynchronously.
+            window.setTimeout(async () => {
+                const promise = targetDiv.requestPointerLock();
+                targetDiv.remove();
+                try {
+                    await promise;
+                    testFailed("Promise resolved despite element removal.");
+                } catch (error) {
+                    window.error = error;
+                    shouldBe("error.name", "'WrongDocumentError'");
```

Comment:
Can probably use shouldBeEqualToString here:

```suggestion
                    shouldBeEqualToString("error.name", "WrongDocumentError");
```

Reply:
I ended up having to change it a bit anyway to deal with a behavior difference between WK1 and WK2. I didn't want the test to test implementation differences that aren't covered by the spec, and I didn't really want to have separate test expectations for WK1 and WK2, so I changed the test like so:

```javascript
const promise = targetDiv.requestPointerLock();
targetDiv.remove();
try {
    await promise;
    // depending on the implementation, requestPointerLock may be satisfied synchronously (WK1).
    // This is OK, but now at this point the pointer should be unlocked.
} catch (error) {
    // If an error does arise due to async pending pointer lock, it should be WrongDocumentError 
    // because the element lost its document that needed to be active and focused.
    if (error.name != "WrongDocumentError")
        testFailed(`Promise rejected with unexpected error ${error}`);
}
shouldBeNull("document.pointerLockElement");
finishJSTest();
```

The reason I used `if (error.name ...` instead of shouldBeEqualToString was to avoid putting a line in the output that is conditional on WK2 vs WK1.

Alternative solutions I can see are:

1. Make WK1 tests behave like WK2 by changing TestController's [requestPointerLock](https://github.com/WebKit/WebKit/blob/b6b74556b8e6ecc165a8e1edb5fc73cc8e5853c2/Tools/WebKitTestRunner/TestController.cpp#L310) to be:

```cpp
static void requestPointerLock(WKPageRef page, const void*)
{
    RunLoop::main().dispatch([=] {
        WKPageDidAllowPointerLock(page);
    });
}
```


2. Define different test expectations for WK1 vs WK2.

3. Punt on this test entirely.

Reply:
Would it be safe to skip WK1? Or maybe yeah, option 1? @smfr?

Reply:
Another option is in Chromium, they have a `testRunner.setPointerLockWillRespondAsynchronously()` method they can call. You can see the vestige of this in [this test](https://github.com/WebKit/WebKit/blob/9a518a5530263ef6ccdce8706fccd4ae5e4c9f2e/LayoutTests/pointer-lock/pointerlockelement-null-when-pending.html#L20) that's disabled in WebKit. I actually implemented that locally in WebKit when I was writing this test, but abandoned it when I found out I could get by without it. I could bring it back if you think that's the best solution here, though.